controller/api/destination/watcher/endpoints_watcher.go

```diff
@@ -24,19 +25,45 @@ const (
 // https://github.com/linkerd/linkerd2/issues/2204
 
 type (
-	// Address represents an individual port on an specific pod.
-	Address struct {
+
+	// Address represents an individual port on a specific endpoint.
+	// This endpoint might be associated with a Pod, a Headless service
+	// or a mirrored service that routes to a remote gateway.
+	Address interface {
```

Comment:
There's nothing I love more than some good old fashioned polymorphism.  However, I'm not sure it's a good fit here.  This interface gives a false sense of encapsulation: we can use this interface in some places, but in others we have to reach inside and switch on the concrete type.

It may be simpler to deal with this as a single struct.  The `Pod`, `OwnerName`, and `OwnerKind` fields can simply be empty when they don't apply (i.e. when the address does not correspond to a pod) just like today.  And we can add an `Identity` field which can provide the identity name when the pod is not present.  I would perhaps call `FullyQualifiedRemoteName` something more aligned with how its effect such as `AuthorityOverride` or something like that.  Alternatively, we can omit this field for now since it's unused and a better name might be more obvious once we have the destination API change solidified.  The `Namespace` and `LocalServiceName` fields appear to be completely unused.  If the goal is to use them to propagate metric labels, we may want to simply have a labels map on the Address struct that we can set directly.

Reply:
> This interface gives a false sense of encapsulation: we can use this interface in some places, but in others we have to reach inside and switch on the concrete type.

Maybe its me but I do not see this as such a problem. Seems in line with trying to express the kind of addresses as a collection of ADTs. 

> t may be simpler to deal with this as a single struct. The Pod, OwnerName, and OwnerKind fields can simply be empty when they don't apply (i.e. when the address does not correspond to a pod) just like today. 

The thing I did not like about this approach (as this is how it was before) is that we are using the presence or abence of values to determine what _type_ of thing we are lookin at. Is this a headless service or a pod address, or whatever else.. 

That all being said, lets try and see how it looks without the different types.

Reply:
This gets a bit philosophical, but I don't think the presence or absence of the pod field fundamentally changes the type of the address.  The way I think of it is this: if the Address has a pod, we'll use that to populate certain metadata; if it doesn't we won't.

Reply:
putting it that way, it makes sense.